### PR TITLE
fix(taskframework): job is not started  in resource idle node when deployed multi-nodes

### PR DESCRIPTION
**Pull request by krihy**
#### What type of PR is this?

type-bug

#### What this PR does / why we need it:
job is not started  when multi nodes has enough resource,  when JobStore acquire trigger but resource is not available , we should release it  and give a chance to other node acquire trigger.

#### Special notes for your reviewer:

test case with two nodes, task be started in equalization in two nodes.
image

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/TaskFrameworkConfiguration.java [hunk 1]
```diff
@@ -84,6 +86,9 @@ public SchedulerFactoryBean taskFrameworkSchedulerFactoryBean(DataSource dataSou
         schedulerFactoryBean.setSchedulerName(taskFrameworkSchedulerName);
         schedulerFactoryBean.setStartupDelay(taskFrameworkProperties.getQuartzStartDelaySeconds());
         schedulerFactoryBean.setTaskExecutor(executor);
+        Properties properties = new Properties();
+        properties.put("org.quartz.jobStore.class", ResourceDetectJobStore.class.getName());
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/TaskFrameworkConfiguration.java:90]: magic value

Comment (yhilmare) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/TaskFrameworkConfiguration.java:90]: is it a standard method to set custom JobStore?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/ResourceDetectJobStore.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            }
+        }
+        return filteredTriggers;
+    }
+
+    private boolean shouldAcquire(OperableTrigger trigger) {
+        TaskFrameworkProperties taskFrameworkProperties = TaskFrameworkPropertiesSupplier.getSupplier().get();
+        // If resource is not available return false and give a chance to other node acquire trigger
+        if (!ResourceDetectUtil.isResourceAvailable(taskFrameworkProperties)) {
+            if (TriggerKey.triggerKey("startPreparingJob", JobConstants.ODC_JOB_MONITORING)
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/ResourceDetectJobStore.java:64]: what does "startPreparingJob" means? why is it this value?

**Comment by yhilmare** [comment]
I can not accept your solution:

1. StartPreparingJob is a daemon job on every node, we'd better not disable it on a specific node
2. we have already got rateLimiter `MonitorProcessRateLimiter`, why does it work?
3. Will other nodes be affected if you `releaseAcquiredTrigger` in `ResourceDetectJobStore`?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/TaskFrameworkConfiguration.java (hunk 1 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/TaskFrameworkConfiguration.java:90]: yes,  it's standard method,  reference `SchedulerFactoryBean` line 575 ,
`mergedProps.putIfAbsent(StdSchedulerFactory.PROP_JOB_STORE_CLASS, LocalDataSourceJobStore.class.getName());`
but the magic value can be replaced by constants

**Comment by krihy** [comment]
> I can not accept your solution:
> 
> 1. StartPreparingJob is a daemon job on every node, we'd better not disable it on a specific node
> 2. we have already got rateLimiter `MonitorProcessRateLimiter`, why does it work?
> 3. Will other nodes be affected if you `releaseAcquiredTrigger` in `ResourceDetectJobStore`?

in cluster model,  all quartz node seize trigger, and be fired exact only once,  if one node  with no resource acquire trigger and fired it, the job will not be started in `StartPreparingJob` in `QuartzSchedulerThread` while loop  and  other node with enough resource will not acquire trigger

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/ResourceDetectJobStore.java (hunk 2 above)
Comment (krihy) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/task/config/ResourceDetectJobStore.java:64]: `JobStore` acquire one trigger at one time, so we should identify the trigger name, and detect current node resource is enough

**Comment by yizhouxw** [comment]
> > I can not accept your solution:
> > 
> > 1. StartPreparingJob is a daemon job on every node, we'd better not disable it on a specific node
> > 2. we have already got rateLimiter `MonitorProcessRateLimiter`, why does it work?
> > 3. Will other nodes be affected if you `releaseAcquiredTrigger` in `ResourceDetectJobStore`?
> 
> in cluster model, all quartz node seize trigger, and be fired exact only once, if one node with no resource acquire trigger and fired it, the job will not be started in `StartPreparingJob` in `QuartzSchedulerThread` while loop and other node with enough resource will not acquire trigger

as we discused before, in cluster model, each node should do StartPreparingJob

**Comment by yhilmare** [comment]
> > I can not accept your solution:
> > 
> > 1. StartPreparingJob is a daemon job on every node, we'd better not disable it on a specific node
> > 2. we have already got rateLimiter `MonitorProcessRateLimiter`, why does it work?
> > 3. Will other nodes be affected if you `releaseAcquiredTrigger` in `ResourceDetectJobStore`?
> 
> in cluster model, all quartz node seize trigger, and be fired exact only once, if one node with no resource acquire trigger and fired it, the job will not be started in `StartPreparingJob` in `QuartzSchedulerThread` while loop and other node with enough resource will not acquire trigger



> > I can not accept your solution:
> > 
> > 1. StartPreparingJob is a daemon job on every node, we'd better not disable it on a specific node
> > 2. we have already got rateLimiter `MonitorProcessRateLimiter`, why does it work?
> > 3. Will other nodes be affected if you `releaseAcquiredTrigger` in `ResourceDetectJobStore`?
> 
> in cluster model, all quartz node seize trigger, and be fired exact only once, if one node with no resource acquire trigger and fired it, the job will not be started in `StartPreparingJob` in `QuartzSchedulerThread` while loop and other node with enough resource will not acquire trigger

so, why doesn't we rely on `MonitorProcessRateLimiter` but disable the `StartPreparingJob`?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/task/constants/JobConstants.java [hunk 3]
```diff
@@ -59,6 +59,8 @@ public class JobConstants {
 
     public static final String ODC_JOB_MONITORING = "odcJobMonitoring";
 
+    public static final String START_PREPARING_JOB = "startPreparingJob";
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/task/constants/JobConstants.java:62]: StdJobScheduler#initStartPreparingJob may use this constants

**Pull request closed without merging**